### PR TITLE
[RFC] Extend Threat Fieldset - Stage 0 Proposal

### PR DESCRIPTION
**Pull request by peasead**
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? Yes
- Have you followed the [contributor guidelines](https://github.com/elastic/ecs/blob/master/CONTRIBUTING.md)? Yes
- For proposing substantial changes or additions to the schema, have you reviewed the [RFC process](https://github.com/elastic/ecs/blob/master/rfcs/README.md)? Yes
- If submitting code/script changes, have you verified all tests pass locally using `make test`? Yes
- If submitting schema/fields updates, have you generated new artifacts by running `make` and committed those changes? Yes
- Is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed. Yes
- Have you added an entry to the [CHANGELOG.next.md](https://github.com/elastic/ecs/blob/master/CHANGELOG.next.md)? Yes

[Preview of the markdown proposal](https://github.com/elastic/ecs/blob/f37d1f78455bf6f589d437ef83cc943ad7b655b7/rfcs/text/0000-extend-threat-group-software.md)

### CODE REVIEW

File: rfcs/text/0000-extend-threat-group-software.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+<!--
+Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting.
+-->
+
+<!--
+Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
+-->
+
+There are many sources of threat intelligence including open source, closed source, and membership-based ISAC's. Depending on the source the level of details can vary from atomic indicators of compromise (IoC) to higher-order context around threat tactics, infrastructure, and motivations. Generally freely available (open source) intelligence sources will provide details more focused on indicators and commercial intelligence services will provide higher-order details.
```
Comment (ebeahan) [inline rfcs/text/0000-extend-threat-group-software.md:98]: Is this a remnant from https://github.com/elastic/ecs/blob/master/rfcs/text/0008-threat-intel.md#source-data that was left in?

I felt a bit of déjà vu, and then I realized why! 😄

File: rfcs/text/0000-extend-threat-group-software.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+<!--
+As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
+Feel free to remove these comments as you go along.
+-->
+
+<!--
+Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
+-->
+
+Currently the `threat` fieldset includes tactic, technique, and sub-techniques from the ATT&CK framework. ATT&CK also includes groups and software that we can easily add to the existing fieldset to include all of the ATT&CK framework. While these fields are directly referenced within the ATT&CK framework, they can also be used with other frameworks if `threat.framework` expands to use more than ATT&CK.
```
Comment (ebeahan) [inline rfcs/text/0000-extend-threat-group-software.md:16]: Thanks for including the consideration for other frameworks. `threat.*` is certainly very much aligned with ATT&CK today, but we don't want it to come off as _exclusively_ ATT&CK focused.

File: rfcs/text/0000/threat.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+- name: threat
+
+  fields:
+
+  - name: software.id
+    level: extended
+    type: keyword
+    short: Software ID assigned by MITRE ATT&CK.
+    description: >
+      Software ID assigned by MITRE for this code, OS utility, open source software, or tool used to conduct behavior modeled in ATT&CK.
```
Comment (ebeahan) [inline rfcs/text/0000/threat.yml:11]: Doesn't need to be addressed for the stage 0 review, but there are conventions for MITRE ATT&CK® that we'll also want to follow in these descriptions.

You can see a list in https://github.com/elastic/ecs/pull/866.

**Review by ebeahan — DISMISSED** [review]
LGTM for stage 0 👍

**Review by ebeahan — APPROVED** [review]

File: rfcs/text/0000-extend-threat-group-software.md (hunk 2 above)
Comment (devonakerr) [reply rfcs/text/0000-extend-threat-group-software.md:16]: Agreed, ATT&CK may currently be the contemporary favorite but we should be mindful that a) it isn't universally relied upon and b) it's not permanent.

File: rfcs/text/0000-extend-threat-group-software.md (hunk 1 above)
Comment (peasead) [reply rfcs/text/0000-extend-threat-group-software.md:98]: Yeah, I can remove this.

File: rfcs/text/0000-extend-threat-group-software.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+<!--
+Stage 2: Add or update all remaining field definitions. The list should now be exhaustive. The goal here is to validate the technical details of all remaining fields and to provide a basis for releasing these field definitions as beta in the schema. Use GitHub code blocks with yml syntax formatting.
+-->
+
+### Proposed New Fields for Threat fieldset
+
+Field | Type | Example | Description
+--- | --- | --- | ---
+threat.software.id | keyword | S0552 | Software is a generic term for custom or commercial code, operating system utilities, open-source software, or other tools used to conduct behavior modeled in ATT&CK. This is the software identification number assigned by MITRE ATT&CK.
```
Comment (devonakerr) [inline rfcs/text/0000-extend-threat-group-software.md:32]: Andy, given that we acknowledge ATT&CK may not be the only/permanent attack model, is there language we should add to indicate that these are just ATT&CK-specific examples? Maybe we could instead use something like "Software is a generic term for custom or commercial code...or other tools used to conduct behavior, commonly modeled using ATT&CK. This value is used for software identification, an example has been provided using ATT&CK."

Comment (peasead) [reply rfcs/text/0000-extend-threat-group-software.md:32]: Yeah, absolutely I can do that. 

TBH, I think that we should drop all of this under `threat.framework.att&ck.*` but that'll be a breaking change that will be as impactful if we do it now or later...so I think we can spend more time now getting everything prepared then making sweeping changes when we're ready.

Comment (devonakerr) [reply rfcs/text/0000-extend-threat-group-software.md:32]: Ah, so you're imagining a situation where there might be other threat.framework definitions - that strikes me as sound, though I couldn't name the top 3-4 alternatives that have a robust knowledge base.

File: rfcs/text/0000/threat.yml (hunk 3 above)
Comment (peasead) [reply rfcs/text/0000/threat.yml:11]: Cool. I made a few updates and we can tighten it up as we move forward.

File: rfcs/text/0000-extend-threat-group-software.md (hunk 4 above)
Comment (peasead) [reply rfcs/text/0000-extend-threat-group-software.md:32]: I completely agree...I'm just trying not to box us into ATT&CK this top level.

I think we're going to get to a point that we may need to next things under `threat.framework.att&ck` (or the like)
Example
```
threat.framework.att&ck.tactic.id
threat.framework.att&ck.tactic.name
threat.framework.att&ck.technique.id
threat.framework.att&ck.technique.name
threat.framework.att&ck.software.id
threat.framework.att&ck.software.name
...
```

I think that'll be a big change though.

**Review by devonakerr — DISMISSED** [review]
Looks good, for the sake of posterity (and perhaps an eye on the horizon) someday we may want to update the Threat RFC's ATT&CK commentary to reflect ATT&CK-defined elements vs calling them ATT&CK elements (ex: ATT&CK-defined software vs ATT&CK software). It's a very minor point, and one which is unlikely to confuse the user, so no change recommended or required.

**Comment by ebeahan** [comment]
Thanks, all! I'll set the advance date, RFC #, and merge it! 🚢

**Review by ebeahan — APPROVED** [review]

**Pull request merged**